src/expr/src/vector_op/jsonb_concat.rs

```diff
@@ -0,0 +1,95 @@
+// Copyright 2023 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use risingwave_common::types::{JsonbRef, JsonbVal};
+use risingwave_expr_macro::function;
+use serde_json::{json, Value};
+
+/// Concatenates the two jsonbs.
+///
+/// Examples:
+///
+/// ```slt
+/// # concat
+/// query T
+/// SELECT '[1,2]'::jsonb || '[3,4]'::jsonb;
+/// ----
+/// [1, 2, 3, 4]
+///
+/// query T
+/// SELECT '{"a": 1}'::jsonb || '{"b": 2}'::jsonb;
+/// ----
+/// {"a": 1, "b": 2}
+///
+/// query T
+/// SELECT '[1,2]'::jsonb || '{"a": 1}'::jsonb;
+/// ----
+/// [1, 2, {"a": 1}]
+///
+/// query T
+/// SELECT '1'::jsonb || '2'::jsonb;
+/// ----
+/// [1, 2]
+///
+/// query T
+/// SELECT '[1,2]'::jsonb || NULL::jsonb;
+/// ----
+/// null
+///
+/// query T
+/// SELECT NULL::jsonb || '[1,2]'::jsonb;
+/// ----
+/// null
+/// ```
+#[function("jsonb_cat(jsonb, jsonb) -> jsonb")]
+pub fn jsonb_cat(left: Option<JsonbRef<'_>>, right: Option<JsonbRef<'_>>) -> JsonbVal {
+    let left_val = left.map_or(Value::Null, |v| v.value().clone());
+    let right_val = right.map_or(Value::Null, |v| v.value().clone());
```

Comment:
SQL `null` is different from JSON `null`. The following is obtained from PostgreSQL:
```
test=# with t(v) as (values (null::jsonb), ('null'::jsonb)) select
test-#     v,
test-#     v is null,
test-#     jsonb_typeof(v),
test-#     v || '[1, 2]'::jsonb
test-# from t;
  v   | ?column? | jsonb_typeof |   ?column?   
------+----------+--------------+--------------
      | t        |              | 
 null | f        | null         | [null, 1, 2]
(2 rows)

test=# \pset null sqlnull
Null display is "sqlnull".
test=# with t(v) as (values (null::jsonb), ('null'::jsonb)) select
test-#     v,
test-#     v is null,
test-#     jsonb_typeof(v),
test-#     v || '[1, 2]'::jsonb
test-# from t;
    v    | ?column? | jsonb_typeof |   ?column?   
---------+----------+--------------+--------------
 sqlnull | t        | sqlnull      | sqlnull
 null    | f        | null         | [null, 1, 2]
(2 rows)
```

In the `slt` tests, we set sql null display as `null` rather than the default empty string. This helped us disambiguate it from an empty varchar output, but may be causing confusion here for jsonb.

Talking about this jsonb concat operation:
* jsonb null does not need any special treatment. It is just a non-array non-object value like `1`
* sql null can be handled by the expr macro behind the scene by removing `Option` from parameters. This means it would always return sql null when any of inputs is sql null. This is called a strict function in PostgreSQL:
```
test=# select proname, proisstrict from pg_proc where proname = 'jsonb_concat';
   proname    | proisstrict 
--------------+-------------
 jsonb_concat | t
(1 row)
```